keyboard: avoid stuck keys due to keybindings (alternate approach)

jlindgren90 opened the pull request:
For discussion/consideration along with #1212 (not mutually exclusive) for solving #1208.

Before commit e77330bc3fe7, there were issues with keys becoming "stuck" if other keys were pressed at the time a keybinding was matched, because those other keys were included in the "bound" set and the release events were incorrectly eaten by labwc.

Commit e77330bc3fe7 solved that issue with the "big hammer" approach of preventing keybindings from working at all if other keys were pressed:

        if (key_state_nr_pressed_keys() > 1) {
                return false;
        }

This is an alternate approach to solving the original problem, by (1) not including those other keys in the "bound" set and (2) making sure we always forward release events for un-bound keys to clients (even if a menu or OSD is displayed).

Details:

- Since we only ever want to store the single matched keycode as bound, `key_state_store_pressed_keys_as_bound()` doesn't really make sense in the plural, so rename it to `key_state_store_pressed_key_as_bound()` and pass in the keycode.

- The calls to `key_state_store_pressed_keys_as_bound()` within `handle_keybinding()` appear to be redundant since it is also called from the parent function (`handle_compositor_keybindings()`). So remove these calls.

- Finally, rework the logic for handling key-release events so that we always forward release events for keys not in the "bound" set.

This PR does not remove the `key_state_nr_pressed_keys() > 1` check, and because of that should not result in any functional change. It should however make it possible to relax or remove that check in future.

johanmalm commented:
I'll have to look at this tomorrow.

At a glance, this refactoring makes me think we (I) made a mistake in the "big hammer" commit in that we no longer store modifiers in the "pressed" array, but it's important that we don't pass release events for bound modifiers too (consider Alt in particular).

Consolatis commented:
> At a glance, this refactoring makes me think we (I) made a mistake in the "big hammer" commit

IMHO at this point in the dev cycle its a shared responsibility, especially in regards to our PR merge rules.
E.g. no point in pointing out any single person to blame and thus the `we` sounds about right :)

> in that we no longer store modifiers in the "pressed" array, but it's important that we don't pass release events for bound modifiers too (consider Alt in particular).

Yeah.. thought the same.

Add to the mix that we also receive key press/release events for modifiers as well (which are sometimes not even detectable as modifiers in which case we can't really detect them being part of a keybind we just executed).

Another thing is that we execute keybinds which may cause focus switching (and thus the seat_enter with the set of pressed keys) on `press`, that might also play a role here.

The whole keybind handling is quite of a mess but I am not sure how we can properly design it differently. There might be a good long term solution but I couldn't think of one so far.

Personally I'll take some keybinds not working over repeating keys all the time. When you had a wrong `enter`, `tab` or `del` key fake repeat on a client before you know what I mean. Pretty funny in a file browser for example. Similar but not as bad when switching to a code editor or chat client.

jlindgren90 commented:
> it's important that we don't pass release events for bound modifiers too (consider Alt in particular).

Good point. I assumed we wanted to pass those through, but in testing with Openbox, it appears not to (i.e. Firefox menus don't appear if you Alt-Tab and cycle back to Firefox).

> Personally I'll take some keybinds not working over wrongly repeating keys all the time.

For me, the issue is that all keybindings will stop working at random times. That's a pretty big usability issue to me, and I think we need to find a way to solve both that and the wrongly repeating keys.

jlindgren90 commented:
> which are sometimes not even detectable as modifiers in which case we can't really detect them being part of a keybind we just executed

Could you explain this part a little more? Is it talking about odd modifiers like `Fn`?

johanmalm commented:
So, maybe this patch is good if we just add currently pressed modifiers in `key_state_store_pressed_key_as_bound()`.

And... the modifiers are held in `seat->keyboard_group->keyboard->modifiers` (see `seat.c+498`)

johanmalm commented:
...and we remove the `key_state_nr_pressed_keys() > 1` hammer

jlindgren90 commented:
> So, maybe this patch is good if we just add currently pressed modifiers in key_state_store_pressed_key_as_bound().

All currently pressed modifiers, or just those that are part of the keybinding?

Consolatis commented:
> > which are sometimes not even detectable as modifiers in which case we can't really detect them being part of a keybind we just executed
> 
> Could you explain this part a little more? Is it talking about odd modifiers like `Fn`?

See #1202 "nitty gritty details" [0]
It might be similar for the `fn` key, that could explain #1223 but I didn't test that.

We may also get different keysyms (and thus know if something is modifier or not) depending on the ordering the keys are pressed and released.

johanmalm commented:
> All currently pressed modifiers, or just those that are part of the keybinding?

I think those two things have to be the same don't they. 

If "other" modifiers are pressed a keybind shouldn't be valid.

For example, if we have defined C+a as a keybind, then C+W+a should NOT trigger it.

It's only the non-modifier bit that has to be "only one" (and the presence of others such non-modifiers should ideally disqualify the bind, but it looks like reality is messy and that we should probably give up on that)

jlindgren90 commented:
Okay that makes sense. I was thinking about modifiers like Num Lock but those aren't really "pressed", so probably irrelevant in this case.

jlindgren90 commented:
For the ISO_Next_Group thing (translated Alt or Shift), as long as we clear both pressed+bound for the correct (raw) key-code we should be fine, I think? Assuming that ISO_Next_Group is not itself part of a keybinding, we will forward the release event to clients, which I think is no different than what happens today.

Consolatis commented:
> Assuming that ISO_Next_Group is not itself part of a keybinding, we will forward the release event to clients, which I think is no different than what happens today.

Well, it might be. A user could configure `Super_L` to be the trigger for switching the keyboard layout. That would then cause it to be recognized as `ISO_Next_Group`. The user could then define any keybind that involves `W-`  like `W-t` or `W-C-t`.

jlindgren90 commented:
In that case the concern would be that we forward an unexpected key release event (Super_L) after the keybinding because we didn't originally recognize the key press as a modifier -- do I understand correctly?

What should we (ideally) do in that case?

johanmalm commented:
> In that case the concern would be that we forward an unexpected key release event (Super_L) after the keybinding because we didn't originally recognize the key press as a modifier -- do I understand correctly?
> 
> What should we (ideally) do in that case?

I’ve not tested, but suspect we have to accept that that whatever key or key-combo is used to switch keyboard layout will not work well (or at all) for the compositor keybinds. I’m okay to accept that.

Whilst of course related, I’d be okay with thinking about that separately.
It’s already an issue today… for example if alt+shift is used for layout switch, then alt+shift+tab doesn’t work.

Consolatis commented:
I currently do not have the brain capacity to wrap my head around the input handling.
Instead, I'd want to point out that we are planning to release [0.6.6](https://github.com/labwc/labwc/issues/1220) in about a month.

So some generic thoughts instead:
- I think there may be long term improvements to make our key handling more robust
- We have a current issue in that some 'stuck' key prevent all further keybinds to work
- Personally, I really don't want fake repeating keys to return (e.g. sending a `press` to a client but eat up the `release`)
- Sending `release` to a client without having sent a `press` first also caused issues in the past
- IMHO for now we should focus on either working around (safety net) or fixing the current issue regarding restoring the keybinds on 'stuck' keys
- We should try not to do too much generic changes regarding input handling until after the release. This part of the codebase is really finicky in that one may change one thing which improves 90% of the usual usage cases but also introduces hard to debug bugs in another area.

If possible I'd like to leave this issue to the both of you, feel free to ping me for specific input though.

johanmalm commented:
^ Thanks @Consolatis I'll try to find a clear slot (rare commodity!) over the next week to try to think this through and propose a patch to get us to a 'safe' state. Having had a finger in the keyboard-pie for a long time, I think I can remember most of the pitfalls. But yeah, I agree - this really is a tough area to get right.

jlindgren90 commented:
I'm all for a targeted/conservative fix if that exists. Not sure if it does but looking forward to what @johanmalm comes up with.

Currently my free time is getting sucked up by home improvement projects, so it might be a little longer before I can get back to looking at this.

johanmalm commented:
Hey, I'll have a go, but don't get hopes up too much 😄 
The last couple of patches I did on keyboard.c turned out not to be too clever when they got some eyeballs on them!

jlindgren90 commented:
I had a little more time to work on this, and I think I addressed the comments so far. Firefox no longer displays its menu bar after Alt-Tab. I'll continue to test and see if any other issues crop up.

johanmalm commented on src/input/key-state.c:
```diff
@@ -69,20 +69,32 @@ key_state_nr_pressed_sent_keycodes(void)
 {
-	memcpy(bound.keys, pressed.keys, MAX_PRESSED_KEYS * sizeof(uint32_t));
-	bound.nr_keys = pressed.nr_keys;
+	add_key(&bound, keycode);
```
In the …bound() function do we need to nullify the array before adding?

bound.nr_keys = 0

jlindgren90 replied:
I think it doesn't matter one way or the other. If there are still pressed/held modifiers from a previous keybinding (that haven't been released yet) then we will add them a second time, but IIRC add_key() has a check to avoid adding a duplicate key-code.

jlindgren90 replied:
Actually, thinking about this more, I think we do not want to clear/nullify the bound array before adding. There may be a non-modifier key still pressed from a previous keybinding (if you pressed two in quick succession) and that should still be considered "bound".

johanmalm replied:
When I made the comment, I had forgotten that there is a `key_state_bound_key_remove()`, so thought we just kept adding :smile:

You make a good point. Agree.

johanmalm reviewed (commented):
I think this is the right direction, i.e. make …bound() function singular and add modifiers when called (like you’ve done).

Have just had look on phone. Just one comment re nullifying array.

Couldn’t follow logic change in first commit (but probably just me)

Will review on computer and test later.

jlindgren90 commented:
The first commit is probably less readable than it could be due to the indentation change. I could rework it a little to make the diff smaller if that's helpful.

Consolatis commented on src/input/key-state.c:
```diff
@@ -69,20 +69,32 @@ key_state_nr_pressed_sent_keycodes(void)
-	memcpy(bound.keys, pressed.keys, MAX_PRESSED_KEYS * sizeof(uint32_t));
-	bound.nr_keys = pressed.nr_keys;
+	add_key(&bound, keycode);
+	/*
+	 * Also store any pressed modifiers as bound. This prevents
+	 * applications from seeing and handling the release event for
+	 * a modifier key that was part of a keybinding (e.g. Firefox
+	 * displays its menu bar for a lone Alt press + release).
+	 */
+	for (int i = 0; i < pressed_mods.nr_keys; ++i) {
+		add_key(&bound, pressed_mods.keys[i]);
+	}
```
Just wanted to point out something: In a `W-Left` example,
- We'd send a press event for the `Super` key to the focused surfaces (`Super_L` or `Super_R`).
- Once the `m` comes in we'd then handle they keybind
- Store the currently set of pressed keys and modifiers as bound
- And thus don't send the release for the earlier `Super_L` / `Super_R`, resulting in the client getting a stray `Press` for the modifier key itself

We can't really prevent the `Super_L` / `Super_R` key to be sent to the application because at that point we don't yet know if it is actually a part of a keybind or not. But I guess we should try to make sure not the eat its release *if* it has already been sent. Maybe applications are also fine with `Press` events for modifiers, I don't know. It feels wrong though unless I miss something.

As we currently do not store the modifiers as pressed this would not happen and the application would see the corresponding `Release` event.

Another thing: the `is_modifier` state may be wrong in case of layout switching, e.g. it might show up as that group_next thing and thus end up in the usual pressed array, depending on the order of Press and Release events for the layout switch.

johanmalm replied:
> We can't really prevent the `Super_L` / `Super_R` key to be sent to the application because at that point we don't yet know if it is actually a part of a keybind or not. But I guess we should try to make sure not the eat its release _if_ it has already been sent. Maybe applications are also fine with `Press` events for modifiers, I don't know. It feels wrong though unless I miss something.

I've never seen this documented anywhere, but believe applications cope fine with seeing modifier press without associated release. And we've done it like that for a long time. I think `alt` is the example the convinced me... (press+release triggers menu). IIRC the thing the makes applications misbehave is sending a release without having first sent a press.
 
> As we currently do not store the modifiers as pressed this would not happen and the application would see the corresponding `Release` event.

Well - until recently they used all (modifiers + non-modifiers) be stored in the pressed array :smile:

> Another thing: the `is_modifier` state may be wrong in case of layout switching, e.g. it might show up as that group_next thing and thus end up in the usual pressed array, depending on the order of Press and Release events for the layout switch.

I tried to reproduce that, but seems to have gone away (i.e. press shift + press caps + release shift + release caps).

I can't test the keybind with `XF86LaunchA` thing (issue #1223). @jp7677 any chance of testing this branch against that?

jlindgren90 replied:
> We can't really prevent the Super_L / Super_R key to be sent to the application because at that point we don't yet know if it is actually a part of a keybind or not. But I guess we should try to make sure not to eat its release if it has already been sent. Maybe applications are also fine with stray Press events for modifiers, I don't know. It feels wrong though unless I miss something.

This initially felt wrong to me too. But it's existing behavior (in both 0.6.5 and master) at least in the Alt-Tab case -- we eat the Alt release-event. And it seems to be the only way to prevent e.g. Firefox from seeing a lone Alt press + release and displaying its menu bar.

> Another thing: the is_modifier state may be wrong in case of layout switching, e.g. it might show up as that group_next thing and thus end up in the usual pressed array, depending on the order of Press and Release events for the layout switch. The corresponding Release may then show up correctly as a modifier. Or the other way around.

This should be handled reasonably well. Both modifiers and non-modifiers are stored in the "pressed" array unconditionally (as raw keycodes). And on release, keycodes are removed from both "pressed" and "pressed_mods" unconditionally. So the only consequence would be in the case where the press event is not detected as a modifier -- then it will not be stored in the "bound" array. I don't think there is much we can do about that.

johanmalm commented:
> The first commit is probably less readable than it could be due to the indentation change. I could rework it a little to make the diff smaller if that's helpful.

I've got my head around it now - and am fine with it.

johanmalm commented:
I've got a good feeling about this patch set. I can't test every case, but I can't reproduce some of the repeat issues we've seen before (e.g. 'other' non-modifier pressed when keybind is processed and getting `...llllllll` after using `W-l` for swaylock).

The logic looks clean and simplifies things in my mind.

I suggest we go with it and see if @kyak + @consus + @jp7677 could try to re-produce recent keyboard issues.

jp7677 commented:
I've tested this branch briefly and the `fn`+`f` issues I reported earlier are solved with this PR, all key bindings with multimedia keys do work.

jlindgren90 commented:
I'd like to figure out what's causing #1238 before we merge this. It may be completely unrelated -- I'm not sure yet.

jlindgren90 commented:
Rebased onto master (no other changes)

kyak commented:
> I've got a good feeling about this patch set. I can't test every case, but I can't reproduce some of the repeat issues we've seen before (e.g. 'other' non-modifier pressed when keybind is processed and getting `...llllllll` after using `W-l` for swaylock).
> 
> The logic looks clean and simplifies things in my mind.
> 
> I suggest we go with it and see if @kyak + @Consus + @jp7677 could try to re-produce recent keyboard issues.

I haven't seen stuck keys in a while; and I also don't reproduce stuck keys with this branch.

johanmalm commented:
Thanks @kyak and @jp7677 for extra testing.
And to @jlindgren90 for persevering with this to get us to a position of cleaner, simpler code and (hopefully) a solid key-handling base.
